ci: attempt to increase stability of saucelabs legacy job

devversion opened the pull request:
See individual commits

devversion committed:
ci: attempt to increase stability of saucelabs legacy job

This is an attempt to increase the stability of the Saucelabs legacy
job by using an emulator recommended by the Saucelabs platform
configurator, explicitly specifying the appium server version etc.

gkalpak commented on karma-js.conf.js:
```diff
@@ -151,6 +151,15 @@ module.exports = function(config) {
 
+  if (process.env.CIRCLECI) {
+    conf.plugins.push(require('karma-parallel'));
+    conf.frameworks.push('parallel');
```
The docs say this has to be the first, so you might need to `unshift` instead 🧐

devversion committed:
build: bundle spec files for saucelabs legacy job

Bundle spec files similar to how it is done within the Angular
Components repo. This should simplify the setup and also speed
up the Saucelab job as only a single spec bundle would need to be
downloaded, compared to having to load hundreds of files through the
Saucelabs tunnel.

Also makes a couple of tests more robust with the emulators/and accounts
for ES2015 test runner changes. The tests should be less reluctant to
such build process changes.

Note for reviewers: Some imports have been simplified here. This work
came from Joey's original WIP for this. It's unclear to me whether this
is still needed, but it sounded like this was necessary for the ESBuild
bundling to work. I have robusted the module resolution plugin though,
so I doubt it's still needed. At the same time though: Not worth
reverting/trying as these changes are nice to have anyway!

Co-Authored-By: Joey Perrott <josephperrott@gmail.com>
Co-Authored-By: Paul Gschwendtner <paulgschwendtner@gmail.com>

devversion committed:
fix(core): destroy hooks not set up for `useClass` provider using `forwardRef`

Fixes an issue where destroy hooks are not set up for `useClass` providers
that rely on `forwardRef` for passing the actual class reference.

Currently the destroy hooks are not captured because forward refs are
not resolved for class providers. In ES2015+ this issue becomes even
more critical if arrow-functions are used for

e.g. `useClass: () => forwardRef(..)`.

This is because arrow functions do not have a prototype, unlike in ES5
where the function assigned to `useClass` would always have a prototype.
In ES2015+ a runtime exception as followed is returned (and doesn't give
any useful indication of the forwardRef issue):

```
TypeError: Cannot read properties of undefined (reading 'ngOnDestroy')
  at registerDestroyHooksIfSupported (dist/packages/core/src/render3/di_setup.ts:196:35 <- dist/legacy-test-bundle.spec.js:41508:37)
  at resolveProvider (dist/packages/core/src/render3/di_setup.ts:153:9 <- dist/legacy-test-bundle.spec.js:41485:11)
  at resolveProvider (dist/packages/core/src/render3/di_setup.ts:70:7 <- dist/legacy-test-bundle.spec.js:41445:9)
  at providersResolver (dist/packages/core/src/render3/di_setup.ts:54:5 <- dist/legacy-test-bundle.spec.js:41438:7)
  at Object.definition.providersResolver (dist/packages/core/src/render3/features/providers_feature.ts:48:18 <- dist/legacy-test-bundle.spec.js:41600:16)
  at resolveDirectives (dist/packages/core/src/render3/instructions/shared.ts:1179:40 <- dist/legacy-test-bundle.spec.js:35611:17)
  at elementStartFirstCreatePass (dist/packages/core/src/render3/instructions/element.ts:40:7 <- dist/legacy-test-bundle.spec.js:38432:27)
  at ɵɵelementStart (dist/packages/core/src/render3/instructions/element.ts:87:7 <- dist/legacy-test-bundle.spec.js:38453:43)
  at ɵɵelement (dist/packages/core/src/render3/instructions/element.ts:180:3 <- dist/legacy-test-bundle.spec.js:38510:5)
  at MultipleFormControls_Template (ng:///MultipleFormControls.js:9:9)
```

devversion commented on packages/core/test/render3/providers_helper.ts:
```diff
@@ -0,0 +1,178 @@
+      App, {injector: defs.ngModule ? createInjector(defs.ngModule) : undefined});
+  expect(fixture.html).toEqual('<parent><view-child>view-child</view-child></parent>');
+
+  fixture.destroy();
```
This helper has just been moved from the spec file to a shared helper file. The only addition is this `fixture.destroy` call here; allowing us to also test that destroy hooks are invoked properly.

AndrewKushnir replied:
I'm wondering if we can have an acceptance test instead (the one in the `packages/core/test/acceptance` folder) vs creating a new test that would use hand-written generated code (which we try to avoid in general)?

devversion committed:
test: re-add downlevel decorator transform test

Re-adds the accidentally deleted downlevel decorator transform test.

devversion committed:
fix(compiler-cli): downlevel transform incorrectly extracting constructor parameters for nested classes

The downlevel decorator transform (commonly used in the CLI and other
tooling of the ecosystem for enabling JIT ES2015+), is currently
incorrectly dealing with nested classes.

The transform will accidentally visit nested classes (in a constructor)
multiple times and generate duplicated instances of the `ctorParameters`
fields. This does not sound like an issue at first, but the duplicated
`ctorParameters` fields will miss significant type information that has
been elided by the first visit, resulting in generated code like the
following:

```js
      let MyClass = /* @__PURE__ */ __name(class MyClass extends UpgradeNg1ComponentAdapter {
        constructor(scope, injector3, elementRef) {
        }
      }, "MyClass");
      MyClass.ctorParameters = () => [
        { type: void 0, decorators: [{ type: Inject, args: [$SCOPE] }] },
        { type: Injector },
        { type: ElementRef }
      ];
      MyClass.ctorParameters = () => [
        { type: void 0 }, // <---- NOTE!
        { type: Injector },
        { type: ElementRef }
      ];
```

devversion committed:
test: remove unused router karma legacy setup

Removes unused router karma legacy test setup that
just caused unnecessary confusion.

devversion committed:
ci: increase jasmine default timeout for saucelabs legacy tests

Previously, when we did not bundle tests, we also had the jasmine
default timeout increased to 15000ms. This commit re-introduces this
increased timeout by also maintaining a custom test init file for the
legacy saucelab jobs. This is desirable as all of the Saucelabs test
code is local in the `tools/legacy-saucelabs` folder.

AndrewKushnir commented on packages/compiler-cli/src/transformers/downlevel_decorators_transform/downlevel_decorators_transform.ts:
```diff
@@ -465,9 +467,8 @@ export function getDownlevelDecoratorsTransform(
-      const updated = ts.updateConstructor(
-          ctor, ctor.decorators, ctor.modifiers, newParameters,
-          ts.visitFunctionBody(ctor.body, decoratorDownlevelVisitor, context));
+      const updated =
```
@alxhub could you please help with the review of the compiler-related changes?

AndrewKushnir commented on packages/core/test/render3/providers_es2015_spec.ts:
```diff
@@ -0,0 +1,45 @@
+
+import {expectProvidersScenario} from './providers_helper';
+
+describe('es2015 providers', () => {
```
It'd be great to add a description on why we have special tests for providers + ES2015 (+ similar comment in the packages/core/test/render3/BUILD.bazel file as well).

AndrewKushnir replied:
FYI, I'm ok to get my feedback addressed in a followup PR, plz don't block the submission on my feedback - there is nothing critical.

devversion replied:
As discussed offline: Definitely agreed. Also on it being an acceptance test instead.

If this PR does not make it today, I'll fix this up tomorrow.

AndrewKushnir replied:
@devversion since this change was presubmitted in g3, let's keep everything as is and create a followup PR (so that we don't need to re-run a presubmit)?

AndrewKushnir commented on packages/router/karma.conf.js:
```diff
@@ -1,113 +0,0 @@
-/**
```
Awesome that we don't have a special Karma config here 👍

AndrewKushnir reviewed (approved):
@devversion thanks for fixing the issues and making CI more stable 👍 

The changes look good (just a couple comments), but I'd like to ask @alxhub to take a look at the compiler-related changes as well (related to the downlevelling logic).

josephperrott reviewed (approved):
LGTM

Reviewed-for: dev-infra

gkalpak commented on karma-js.conf.js:
```diff
@@ -151,6 +149,15 @@ module.exports = function(config) {
     set: () => {},
   });
 
+  if (process.env.CIRCLECI) {
```
Nit: Why not `process.env.CI` (which is what we typically use)?

gkalpak replied:
Nevermind, this code was removed in the next commit.

devversion replied:
I should have cleaned this up but didn't realize. This PR really grew too much with all the other necessary fixes, sorry! 😇😢

devversion replied:
I will try to fix this if the PR does not land today.

gkalpak commented on packages/core/test/animation/animations_with_web_animations_integration_spec.ts:
```diff
@@ -249,7 +249,8 @@ describe('animation integration tests using web animations', function() {
               .list .inner {
-                line-height:50px;
+                box-sizing: border-box;
+                height: 50px;
```
OOC, why this change?

devversion replied:
line height behaved flaky in Android 10, and this fixed it

dylhunn commented:
Igor is the only reviewer on fw-benchmarks and is no longer with us, so we can ignore pullapprove bot for now. What target label does this need?

AndrewKushnir commented:
Merge-assistance: this PR is ready for merge. Alex's approval covers the remaining `fw-benchmarks` group as well (Alex will be a part of that group once PR #44304 is merged).

dylhunn commented:
This PR was merged into the repository by commit 0fcf18d5f476cb2be6c89d28185fdf8ab34251f4.
